docs/getting-started/getting-started-with-docker.rst

```diff
@@ -1,36 +1,42 @@
-
-.. note:: This is the first part of a tutorial series where you will create a tool called BAMStats and publish it onto Dockstore.
-
 Getting Started with Docker
 ===========================
 
 Tutorial Goals
 --------------
+This is the first part of our "Getting Started" tutorial series, where we will walk you through the creation of a tool called BAMstats and publish it onto Dockstore. This particular page, "Getting Started with Docker," will go through the basics of Docker and have you develop a Dockerfile for the BAMstats tool.
+
+This tutorial assumes you have basic knowledge of a Unix-like file system, such as working directories, moving between directories on the command line.
 
 -  Learn about Docker
--  Create a Docker image for a real tool
+-  Create a Dockerfile
+-  Use the Dockerfile you created to make a Docker image for a real bioinformatics tool
 -  Create a tag locally
--  Test Docker image locally
+-  Test your Docker image locally
 
 Introduction to Docker
 ----------------------
 
-.. note:: See `Docker Overview <https://docs.docker.com/get-started/overview/>`__ for an excellent overview about Docker.
+.. note:: See `Docker Overview <https://docs.docker.com/get-started/overview/>`__ for an excellent overview of Docker.
+
+There is a good chance that you have heard of containerization. A container is essentially an emulated computer system that bundles up kind of software. It's similar to a virtual machine, but is usually much lighter on system resources, as a container, unlike a VM, does not need to run an extra operating system.
+
+Docker is a well-known type of containerization software. It provides a fast environment for both users and developers. A developer can package software and its dependencies into an Docker image, and then share that Docker image with users. Users who download your Docker image and run it with the Docker program will be able to run the software you packaged without having to handle the installation of your program's prerequisites or anything else; it's ready to go as-is.
 
-Docker is a fantastic tool for creating light-weight containers to run
-your tools. It gives you a fast, VM-like environment for Linux where you
-can automatically install dependencies, make configurations, and setup
-your tool exactly the way you want, just as you would on a "normal"
-Linux host. You can then quickly and easily share these Docker images
-with the world using registries like Quay.io (indexed by Dockstore),
-Docker Hub, and GitLab.
+Docker images are usually shared on registries such as Quay.io, Docker Hub, and GitLab. When a user downloads an image and runs it locally, it creates an editable copy of the image. The editable copy is called a Docker container. You can think of an image as a template, and a container as something made from that template.
```

Comment:
You might wanna run this by Charles or someone else with more Docker experience than me.  I think of the Docker "image" as the bundled-up set of files, and the "container" as the running application spawned from the image within Docker.

Reply:
I did a lot of research into this for the Dockstore dictionary. Here's my current understanding of the guts of Docker... Charles et al, let me know if this is technically correct, because I base multiple explanations here and elsewhere on this:

A Docker image is a read-only executable file. It is essentially a bundled-up filesystem that usually contains a piece of software and requirements of said software. In bioinformatics our Docker images usually use a Docker image of a Linux filesystem + some Linux distro's goodies (such as apt-get) as the base image, and then add layers on top of it. Every RUN, COPY, and ADD instruction in a Dockerfile adds a new layer on top of the base image/previous layers. That being said, the contents of a Docker image do not necessarily package anything **_from_** any particular operating system, but it does need things **_for_** a given operating system. This can be seen with the absolute minimal Docker image possible, [hello-world](https://registry.hub.docker.com/_/hello-world/?tab=tags), which consists of naught but scratch (the absolute most basic of base images) and [a single binary program](https://github.com/docker-library/hello-world/blob/master/amd64/hello-world/hello). In order to run that binary program correctly, there are different versions of the Docker image for different hardware architectures, as well as different host OSes, but even though they can fluctuate greatly in size for different host OSes/archs (the [smallest](https://registry.hub.docker.com/layers/hello-world/library/hello-world/latest/images/sha256-995efde2e81b21d1ea7066aa77a59298a62a9e9fbb4b77f36c189774ec9b1089?context=explore) is less than 3 KB, the [largest](https://registry.hub.docker.com/layers/hello-world/library/hello-world/latest/images/sha256-1e0054d86f2721056c9db5c2322fb8883d185fa1165bca69cb0ac88eeb40be30?context=explore) is over 100 MB) none of them contain much of anything from any particular operating system; they just need a different setup depending on what host OS/arch they are going to be running on, in the same way that [Debian for arm64](https://cdimage.debian.org/debian-cd/current/arm64/iso-cd/) isn't the same size as [Debian for i386](https://cdimage.debian.org/debian-cd/current/i386/iso-cd/).

When it's time to actually run the software package inside a Docker image, the Docker engine creates an editable copy of the Docker image. This editable copy is a Docker container. (My object-oriented Python brain compares this to an object being an instance of a particular class' non-editable-at-runtime template.)

A Docker image:
* cannot be edited, but a new version can be created via a Dockerfile
* typically bundles up a filesystem, but not an entire operating system, yet might bundle components of an operating system
* is created using `docker build` and a Dockerfile via a process that involves building layers on top of each other

A Docker container:
* can be edited in the sense that you can add/delete files from its filesystem while it is being executed
* mounts its own filesystem separate from the host OS, usually based upon the filesystem provided in the Docker image (the user can mount additional stuff into this filesystem from the host OS using the volume/mount commands)
* runs its own processes in such a way that it can only see said processes, not other stuff going on in the host machine, and with unique user ID(s)
* shares the same kernel as the host OS
* does not have a guest OS; what is being run in a Docker container is technically being run "natively" on the host OS

A Docker tag:
* is a mutable reference to a particular Docker image

A Docker digest:
* is an immutable description of a particular Docker image derived from the `docker build` command that created said 
Docker image
* can be used to reference a particular Docker image more precisely than a tag

The Docker engine:
* Is not a hypervisor
* However, it can manage several Docker containers at the same time, like how a hypervisor can manage several guest OSes at the same time

Reply:
I'm more or less ok with most of this except that "editable copy" is kind of an unusual way of putting it. While true, it seems incomplete? https://docs.docker.com/get-started/overview/#containers has "runnable instance"  which is more familiar to me

Reply:
Given the fact that a Docker image is read-only, I like mentioning that it is editable, but honestly my OOP brain just thinks of it as an instance. Think I'll reword the paragraph this shows up in slightly to get those points across.